### ISSUE TITLE
Remove Guava to make logbook lightweight

### ISSUE THREAD

**Issue by ouaibsky** [zalando/logbook#59]
Hi

I think it's a shame for logbook to get dep on guava which is very huge library
For what I seen, logbook is using java8 and usage of guava in logbook could be replace by basic lambda

If I propose a PR for guava removing, in is there chance to be accepted ?
Thx
Christophe

**Comment by whiskeysierra** [comment]
We are using Guava for much, much more than just functions/predicates. In fact we don't use the functional stuff from guava at all, in favor of the Java 8 versions. Same goes for Optional. We use it for the following:

- Splitter
- Multimap
- ByteArrayDataOutput
- ByteStreams
- CharStreams
- ImmutableCollections
- Preconditions (partially obsolete due to Objects)

None of those has an equivalent in the standard libraries.

You can check for yourself: https://github.com/zalando/logbook/search?utf8=%E2%9C%93&q=com.google

Also I'd like to challenge that
> guava [..] is very huge library

Guava itself has no dependencies, i.e. it's self-contained. Google is actually very reluctant in blindly adding new features to it. We consider it one of the best libraries out there in terms of signal-to-noise ratio and quite frankly, it should be part of the JDK itself.

**Comment by whiskeysierra** [comment]
In addition, if you feel like it, please fork logbook and try to replace all usages of Guava. I'm sure midway through you'll realize that Guava is worth it.

**Comment by ouaibsky** [comment]
Hi

So I was lucky to ask you  this question before starting a PR ;)
My point was not to depreciate guava which is very useful in some case.
The pb (from my point of view, and I accept that it's not shared by everybody) is the size (2.2 Mo) and the fact that you have to pick everything.
Your project weights 18k and aims to be a third library used by more final projects, in some case (no so rare) you have to use many third libraries with conflict in transitive dependencies.

According to me putting this type of dependency is really an architecture choice with a set of trade-off.

I did a quick analysis on logbook-core and removing  guava generates less than 50 compilation errors mainly focused on 
* Multimap (which is mainly a Map<K, Collections<V>)
* ForwardingMultimap
* MediaType and CharMatcher (1 usage)

At the end it's not so much and that's why I though it could be discuss.
But you're the boss on your project :)

Cheers
Christophe.

**Comment by whiskeysierra** [comment]
I wouldn't mind seeing a version of logbook without Guava. If you're
willing to do the work and open a pull request I think we would have more
concrete discussions. I'm not totally opposed to the idea.

**Comment by jbellmann** [comment]
If it's just the size and you just want to remove the unused guava-stuff you could go with the [maven-shade-plugin](https://maven.apache.org/plugins/maven-shade-plugin/). It's just some configuration, keep what you already use, remove the rest. To avoid version-conflicts you can repackage the dependencies you need. But that's something can wait until someone has issues with you library.

**Comment by ouaibsky** [comment]
Yep shape is an option, probably have to repackage to avoid conflict, to be tested
I spent a couple of hours today and closed to have something working.
C.

**Comment by whiskeysierra** [comment]
> If it's just the size and you just want to remove the unused guava-stuff you could go with the maven-shade-plugin.

I just did that locally on a feature branch to get a hang of it.

The setup will probably be a bit tricky to get it ready for production, because multi module + shade usually means we have n copies of our dependency, rather than one, so I guess we need an aggregate, i.e. uberjar of everything (?). Or alternatively we make sure no module depends on guava directly except logbook-api and every other module is relying on the transitive dependency (requires rewriting the pom, iirc).

In terms of size I was able to reduce the size of Guava from 2269kb to 878kb.
